### ISSUE TITLE
`Nexus::instance_start` doesn't check that the target instance is fully created

### ISSUE THREAD

**Issue by gjcolombo** [oxidecomputer/omicron#6274]
This was seen in stress testing of PR #5749. That PR seems to hit the problem more often than main does, for reasons we're still investigating, but it looks like it's also a bug on current main so I'm filing a separate issue for it.

This is a regression I introduced in #4194. `Nexus::instance_start` previously rejected attempts to start a Creating instance: https://github.com/oxidecomputer/omicron/blob/79fee6a2c9f0099fb9cca6effbb862fd75c0a4e4/nexus/src/app/instance.rs#L483-L520

In 4194 I changed this path so that it checks the state of the instance's *VMM* but does not pay attention to the state of the *instance*: https://github.com/oxidecomputer/omicron/blob/acbaf4370b876bd8f27b71bbf41fd12abafc09dc/nexus/src/app/instance.rs#L533-L581

The `if let Some(vmm)` clause should have an `else` that checks to make sure that the instance is actually in a state where it's ready to be started (i.e. it should be in `NoVmm`).

At first glance the impact of this seems pretty limited: to hit it, you need to create an instance, then try to start it by name before the create saga is finished. The result that we're seeing is that the start request 500s; the other possible result, I think, is that the VM will start without all the virtual hardware you might expect, which you can recover from by stopping and restarting the instance.

**Comment by hawkw** [comment]
This issue reproduces on `main` (as of e4067e10db8645667eecd2959ac8dc86330bf7ae) running on  `madrid`:
```logs
2024-08-09T00:22:06.720969Z  WARN actor{name="inst7_3"}:antagonize{instance_name="inst7"}: omicron_stress::actor::instance: 158: instance start request returned result=Err(Error Response: status: 500 Internal Server Error; headers: {"content-type": "application/json", "x-request-id": "61243615-c2bd-434c-b926-e48c3886c24e", "content-length": "124", "date": "Fri, 09 Aug 2024 00:22:06 GMT"}; value: Error { error_code: Some("Internal"), message: "Internal Server Error", request_id: "61243615-c2bd-434c-b926-e48c3886c24e" })
2024-08-09T00:22:06.788471Z ERROR omicron_stress: 191: actor error: Error Response: status: 500 Internal Server Error; headers: {"content-type": "application/json", "x-request-id": "61243615-c2bd-434c-b926-e48c3886c24e", "content-length": "124", "date": "Fri, 09 Aug 2024 00:22:06 GMT"}; value: Error { error_code: Some("Internal"), message: "Internal Server Error", request_id: "61243615-c2bd-434c-b926-e48c3886c24e" }
2024-08-09T00:22:06.788550Z  INFO omicron_stress: 218: Halting actors
```
```console
root@oxz_switch1:~# pilot host exec -O -c '/opt/oxide/oxlog/oxlog zones | grep nexus | xargs -L 1 /opt/oxide/oxlog/oxlog logs --current --archived | xargs -L 1 cat' 0-31 | looker -c 'r.req_id?.contains("61243615-c2bd-434c-b926-e48c3886c24e")' -l info
00:22:06.693Z INFO 25599db0-0d38-4d8b-a2d5-f9606cb19fb8 (dropshot_external): request completed
    error_message_external = Internal Server Error
    error_message_internal = saga ACTION error at node "ensure_registered": Expected exactly one SNAT IP address for an instance
    file = /home/build/.cargo/git/checkouts/dropshot-a4a923d29dccc492/52d900a/dropshot/src/server.rs:902
    latency_us = 25140133
    local_addr = 172.30.2.6:443
    method = POST
    remote_addr = 172.20.17.202:58106
    req_id = 61243615-c2bd-434c-b926-e48c3886c24e
    response_code = 500
    uri = /v1/instances/inst7/start?project=omicron-stress
root@oxz_switch1:~#
```

**Comment by hawkw** [comment]
FWIW, in #5749, I've changed the `InstanceAndVmm::effective_state` method to determine the externally-visible state based on the rules discussed in the "improving instance lifecycle's foundations" document:

https://github.com/oxidecomputer/omicron/blob/2ea3fe637bf24667dcddf31a23a7b4813764d426/nexus/db-queries/src/db/datastore/instance.rs#L88-L174

I've updated many of the API endpoints to use this logic for determining the instance's state. IMO, it's *probably* worth considering reusing that logic when solving this issue as well, so that we don't end up in a situation where the `instance-start` API rejects a start request and claims that the instance is in a particular state, but other APIs suggest it's in a different state. Just a thought.
